File: windows/tests/RealtimeTranslator.Core.Tests/SessionHealthMonitorTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Nodes;
using RealtimeTranslator.Core.Audio;
using RealtimeTranslator.Core.OpenAI;
using RealtimeTranslator.Core.Realtime;
using Xunit;

namespace RealtimeTranslator.Core.Tests;

/// <summary>
/// shared/fixtures/v1/health.json を正本として、SessionHealthMonitor の
/// 位相・検知系列を単調時計上で検証する。
/// </summary>
public sealed class SessionHealthMonitorTests
{
    // Given: health.json の thresholds セクション
    // When: SessionHealthThresholds の既定値を見る
    // Then: fixture の thresholds と一致する
    [Fact]
    public void DefaultThresholdsMatchFixture()
    {
        var thresholds = SharedFixtures.Load("health")["thresholds"]!.AsObject();

        var defaults = new SessionHealthThresholds();
        Assert.Equal(Ms(thresholds, "captureStallMs"), defaults.CaptureStall);
        Assert.Equal(Ms(thresholds, "sendStallMs"), defaults.SendStall);
        Assert.Equal(Ms(thresholds, "receiveStallMs"), defaults.ReceiveStall);
        Assert.Equal(Ms(thresholds, "sourceStallMs"), defaults.SourceStall);
        Assert.Equal(Ms(thresholds, "translationStallMs"), defaults.TranslationStall);
        Assert.Equal(Ms(thresholds, "silenceMs"), defaults.Silence);
        Assert.Equal(Ms(thresholds, "connectGraceMs"), defaults.ConnectGrace);
        Assert.Equal(Ms(thresholds, "expiryNearMs"), defaults.ExpiryNear);
        Assert.Equal(
            thresholds["audioActivityPeakFloor"]!.GetValue<double>(),
            defaults.AudioActivityPeakFloor,
            9);
    }

    // Given: health.json の各 scenario（単調時刻 ms の op 列）
    // When: 時刻順に op を SessionHealthMonitor へ適用する
    // Then: 各 evaluate の phase と新規検知（kind 定義順）が期待と一致する
    [Theory]
    [MemberData(nameof(ScenarioNames))]
    public void ScenarioMatchesFixture(string name)
    {
        var scenario = SharedFixtures.Case("health", "scenarios", name);
        var monitor = new SessionHealthMonitor();

        // steps は op ごとにまとまって並ぶため、単調時刻で安定ソートする。
        var steps = scenario["steps"]!.AsArray()
            .Select((step, index) => (step: step!.AsObject(), index))
            .OrderBy(entry => At(entry.step))
            .ThenBy(entry => entry.index)
            .ToList();

        foreach (var (step, _) in steps)
        {
            var at = TimeSpan.FromMilliseconds(At(step));
            switch (SharedFixtures.Text(step["op"]))
            {
                case "begin":
                    monitor.BeginGeneration(
                        At(step, "generation"),
                        At(step, "epoch"),
                        step["isRecovery"]?.GetValue<bool>() ?? false,
                        at);
                    break;
                case "capture":
                    monitor.RecordCapture(at, step["active"]?.GetValue<bool>() ?? false);
                    break;
                case "send":
                    monitor.RecordSendSuccess(at);
                    break;
                case "receive":
                    monitor.RecordReceive(Lane(step["lane"]), at);
                    break;
                case "source":
                    monitor.RecordSourceProgress(at);
                    break;
                case "translation":
                    monitor.RecordTranslationProgress(Lane(step["lane"]), at);
                    break;
                case "select":
                    monitor.SetSelectedLane(
                        step["lane"] is null ? null : Lane(step["lane"]),
                        at);
                    break;
                case "expiry":
                    monitor.RecordSessionExpiry(
                        Lane(step["lane"]),
                        step["remainingMs"] is { } remaining
                            ? TimeSpan.FromMilliseconds(remaining.GetValue<int>())
                            : null,
                        at);
                    break;
                case "evaluate":
                {
                    var (snapshot, detections) = monitor.Evaluate(at);
                    var expect = step["expect"]!.AsObject();
                    Assert.Equal(
                        SharedFixtures.Text(expect["phase"]),
                        PhaseName(snapshot.Phase));
                    var expected = expect["detections"]!.AsArray()
                        .Select(entry => entry is JsonValue
                            ? (Kind: entry!.GetValue<string>(), Lane: (string?)null)
                            : (
                                Kind: SharedFixtures.Text(entry!["kind"]),
                                Lane: (string?)entry!["lane"]?.GetValue<string>()))
                        .ToList();
                    Assert.Equal(
                        expected.Select(e => e.Kind).ToList(),
                        detections.Select(d => DetectionName(d.Kind)).ToList());
                    // lane が fixture 側で指定されている検知だけ lane を照合する。
                    foreach (var (detection, expectedLane) in detections.Zip(expected))
                    {
                        if (expectedLane.Lane is not null)
                        {
                            Assert.Equal(expectedLane.Lane, detection.Lane?.ToLogString());
                        }
                    }

                    break;
                }
                default:
                    Assert.Fail($"unknown op in {name}: {step}");
                    break;
            }
        }
    }

    // Given: 無音の PCM16 frame
    // When: ピーク振幅を計算する
    // Then: 0 を返す
    [Fact]
    public void Pcm16PeakOfSilenceIsZero()
    {
        Assert.Equal(0, Pcm16AudioActivity.NormalizedPeakAmplitude(new byte[4800]));
    }

    // Given: 1 sample だけ 32767 を含む frame
    // When: ピーク振幅を計算する
    // Then: 1.0 を返す
    [Fact]
    public void Pcm16PeakOfMaxSampleIsOne()
    {
        var frame = new byte[4800];
        frame[100] = 0xFF;
        frame[101] = 0x7F;
        Assert.Equal(1.0, Pcm16AudioActivity.NormalizedPeakAmplitude(frame), 9);
    }

    // Given: Int16 最小値（-32768）だけを含む frame
    // When: ピーク振幅を計算する
    // Then: 飽和して 1.0 を返す
    [Fact]
    public void Pcm16PeakOfInt16MinClampsToOne()
    {
        var frame = new byte[4800];
        frame[0] = 0x00;
        frame[1] = 0x80;
        Assert.Equal(1.0, Pcm16AudioActivity.NormalizedPeakAmplitude(frame), 9);
    }

    // Given: 活動閾値 0.005 前後の振幅を持つ frame
    // When: ピーク振幅を計算する
    // Then: 閾値超過のみが活動とみなされる
    [Fact]
    public void Pcm16PeakAroundActivityFloor()
    {
        var above = new byte[4800];
        above[0] = 0xA4; // 164/32767 ≈ 0.005005
        Assert.True(Pcm16AudioActivity.NormalizedPeakAmplitude(above) > 0.005);

        var below = new byte[4800];
        below[0] = 0x0A; // 10/32767 ≈ 0.0003
        Assert.True(Pcm16AudioActivity.NormalizedPeakAmplitude(below) <= 0.005);
    }

    public static TheoryData<string> ScenarioNames => SharedFixtures.CaseNames("health", "scenarios");

    private static int At(JsonObject step, string field = "at") => step[field]!.GetValue<int>();

    private static TimeSpan Ms(JsonObject node, string field) =>
        TimeSpan.FromMilliseconds(node[field]!.GetValue<int>());

    private static RealtimeTranslationLane Lane(JsonNode? node) =>
        SharedFixtures.Text(node) switch
        {
            "source" => RealtimeTranslationLane.Source,
            "en" => RealtimeTranslationLane.Translation(RealtimeTranslationOutputLanguage.English),
            "ja" => RealtimeTranslationLane.Translation(RealtimeTranslationOutputLanguage.Japanese),
            "es" => RealtimeTranslationLane.Translation(RealtimeTranslationOutputLanguage.Spanish),
            var other => throw new InvalidOperationException($"unknown lane {other}"),
        };

    private static string PhaseName(SessionHealthPhase phase) => phase switch
    {
        SessionHealthPhase.Idle => "idle",
        SessionHealthPhase.CatchingUp => "catchingUp",
        SessionHealthPhase.Silence => "silence",
        SessionHealthPhase.AwaitingLanguageDetection => "awaitingLanguageDetection",
        SessionHealthPhase.Active => "active",
        _ => throw new ArgumentOutOfRangeException(nameof(phase)),
    };

    private static string DetectionName(SessionHealthDetectionKind kind) => kind switch
    {
        SessionHealthDetectionKind.CaptureStalled => "captureStalled",
        SessionHealthDetectionKind.SendStalled => "sendStalled",
        SessionHealthDetectionKind.ReceiveStalled => "receiveStalled",
        SessionHealthDetectionKind.SourceStalled => "sourceStalled",
        SessionHealthDetectionKind.TranslationStalled => "translationStalled",
        SessionHealthDetectionKind.ExpiryNear => "expiryNear",
        SessionHealthDetectionKind.Expired => "expired",
        _ => throw new ArgumentOutOfRangeException(nameof(kind)),
    };
}

File: RealtimeTranslator/Tests/SessionHealthMonitorTests.swift
import XCTest
@testable import RealtimeTranslator

/// shared/fixtures/v1/health.json を正本として、SessionHealthMonitor の
/// 位相・検知系列を単調時計上で検証する。
final class SessionHealthMonitorFixtureTests: XCTestCase {
    // Given: health.json の thresholds セクション
    // When: SessionHealthThresholds の既定値を見る
    // Then: fixture の thresholds と一致する
    func testDefaultThresholdsMatchFixture() throws {
        let root = try SharedFixtures.load("health")
        let thresholds = try XCTUnwrap(root["thresholds"] as? [String: Any])

        let defaults = SessionHealthThresholds()
        XCTAssertEqual(defaults.captureStall, .milliseconds(number(thresholds, "captureStallMs")))
        XCTAssertEqual(defaults.sendStall, .milliseconds(number(thresholds, "sendStallMs")))
        XCTAssertEqual(defaults.receiveStall, .milliseconds(number(thresholds, "receiveStallMs")))
        XCTAssertEqual(defaults.sourceStall, .milliseconds(number(thresholds, "sourceStallMs")))
        XCTAssertEqual(
            defaults.translationStall,
            .milliseconds(number(thresholds, "translationStallMs"))
        )
        XCTAssertEqual(defaults.silence, .milliseconds(number(thresholds, "silenceMs")))
        XCTAssertEqual(defaults.connectGrace, .milliseconds(number(thresholds, "connectGraceMs")))
        XCTAssertEqual(defaults.expiryNear, .milliseconds(number(thresholds, "expiryNearMs")))
        XCTAssertEqual(
            defaults.audioActivityPeakFloor,
            SharedFixtures.real(thresholds["audioActivityPeakFloor"]),
            accuracy: 1e-9
        )
    }

    // Given: health.json の各 scenario（単調時刻 ms の op 列）
    // When: 時刻順に op を SessionHealthMonitor へ適用する
    // Then: 各 evaluate の phase と新規検知（kind 定義順）が期待と一致する
    func testScenariosMatchFixture() throws {
        for name in try SharedFixtures.caseNames("health", "scenarios") {
            let scenario = try SharedFixtures.case("health", "scenarios", name)
            var monitor = SessionHealthMonitor()

            // steps は op ごとにまとまって並ぶため、単調時刻で安定ソートする。
            let rawSteps = try XCTUnwrap(scenario["steps"] as? [[String: Any]])
            let steps = rawSteps.enumerated().sorted {
                number($0.element, "at") == number($1.element, "at")
                    ? $0.offset < $1.offset
                    : number($0.element, "at") < number($1.element, "at")
            }

            for (_, step) in steps {
                let at = Duration.milliseconds(number(step, "at"))
                switch SharedFixtures.text(step["op"]) {
                case "begin":
                    monitor.beginGeneration(
                        generation: number(step, "generation"),
                        epoch: number(step, "epoch"),
                        isRecovery: step["isRecovery"] as? Bool ?? false,
                        now: at
                    )
                case "capture":
                    monitor.recordCapture(now: at, hasAudioActivity: step["active"] as? Bool ?? false)
                case "send":
                    monitor.recordSendSuccess(now: at)
                case "sendStart":
                    monitor.recordSendStart(now: at)
                case "receive":
                    monitor.recordReceive(lane: try lane(step["lane"], stepName: name), now: at)
                case "source":
                    monitor.recordSourceProgress(now: at)
                case "translation":
                    monitor.recordTranslationProgress(
                        lane: try lane(step["lane"], stepName: name),
                        now: at
                    )
                case "select":
                    monitor.setSelectedLane(
                        step["lane"] is NSNull || step["lane"] == nil
                            ? nil
                            : try lane(step["lane"], stepName: name),
                        now: at
                    )
                case "expiry":
                    let remaining: Duration? =
                        step["remainingMs"] is NSNull || step["remainingMs"] == nil
                            ? nil
                            : .milliseconds(number(step, "remainingMs"))
                    monitor.recordSessionExpiry(
                        lane: try lane(step["lane"], stepName: name),
                        remaining: remaining,
                        now: at
                    )
                case "evaluate":
                    let (snapshot, detections) = monitor.evaluate(now: at)
                    let expect = try XCTUnwrap(step["expect"] as? [String: Any])
                    XCTAssertEqual(
                        snapshot.phase.rawValue,
                        SharedFixtures.text(expect["phase"]),
                        "phase mismatch at \(number(step, "at")) in \(name)"
                    )
                    let expectedDetections = try XCTUnwrap(expect["detections"] as? [Any])
                        .map { entry -> (kind: String, lane: String?) in
                            if let text = entry as? String {
                                return (text, nil)
                            }
                            let object = entry as? [String: Any]
                            return (
                                SharedFixtures.text(object?["kind"]),
                                SharedFixtures.optionalText(object?["lane"])
                            )
                        }
                    XCTAssertEqual(
                        detections.map { $0.kind.rawValue },
                        expectedDetections.map { $0.kind },
                        "detection kinds mismatch at \(number(step, "at")) in \(name)"
                    )
                    // lane が fixture 側で指定されている検知だけ lane を照合する。
                    for (detection, expected) in zip(detections, expectedDetections)
                    where expected.lane != nil {
                        XCTAssertEqual(
                            detection.lane?.healthLogName,
                            expected.lane,
                            "detection lane mismatch at \(number(step, "at")) in \(name)"
                        )
                    }
                default:
                    XCTFail("unknown op in \(name): \(step)")
                }
            }
        }
    }

    // Given: en lane に残り 130 秒の期限が記録された monitor
    // When: evaluate の snapshot を description する
    // Then: expiryRemainingMs=en:130000 と sinceCapture/sinceSendSuccess を含む
    func testSnapshotDescriptionIncludesExpiryRemaining() {
        var monitor = SessionHealthMonitor()
        monitor.beginGeneration(generation: 1, epoch: 1, isRecovery: false, now: .zero)
        monitor.recordSessionExpiry(
            lane: .translation(.english),
            remaining: .seconds(130),
            now: .zero
        )

        let (snapshot, _) = monitor.evaluate(now: .zero)
        XCTAssertTrue(snapshot.description.contains("expiryRemainingMs=en:130000"))
        XCTAssertTrue(snapshot.description.contains("sinceCaptureMs=-"))
        XCTAssertTrue(snapshot.description.contains("sinceSendSuccessMs=-"))
    }

    // Given: send が in-flight で capture が止まっている monitor
    // When: captureStall・sendStall の閾値を超える時刻で evaluate する
    // Then: captureStalled は出ず、sendStalled は in-flight 開始から 10s で発火する
    func testInFlightSendSuppressesCaptureStalled() {
        var monitor = SessionHealthMonitor()
        monitor.beginGeneration(generation: 1, epoch: 1, isRecovery: false, now: .zero)
        monitor.recordCapture(now: .milliseconds(100), hasAudioActivity: true)
        monitor.recordSendSuccess(now: .milliseconds(100))
        monitor.recordCapture(now: .milliseconds(200), hasAudioActivity: true)
        monitor.recordSendStart(now: .milliseconds(200))

        // in-flight 開始から 3.1s 後: captureStalled は出ない。
        var (_, detections) = monitor.evaluate(now: .milliseconds(3_300))
        XCTAssertFalse(detections.contains { $0.kind == .captureStalled })

        // in-flight 開始から 10s: sendStalled が発火する。
        (_, detections) = monitor.evaluate(now: .milliseconds(10_200))
        XCTAssertTrue(detections.contains { $0.kind == .sendStalled })
        XCTAssertFalse(detections.contains { $0.kind == .captureStalled })
    }

    private func number(_ object: [String: Any], _ key: String) -> Int {
        SharedFixtures.number(object[key])
    }

    private func lane(_ value: Any?, stepName: String) throws -> RealtimeTranslationLane {
        switch SharedFixtures.text(value) {
        case "source": return .source
        case "en": return .translation(.english)
        case "ja": return .translation(.japanese)
        case "es": return .translation(.spanish)
        default: throw SharedFixtures.FixtureError.invalidCase(
            fixture: "health",
            section: "scenarios",
            name: stepName
        )
        }
    }
}
